### ISSUE TITLE
Function lacks ending return statement and return type does not include 'undefined'

### ISSUE THREAD

**Issue by Mng12345** [microsoft/TypeScript#58099]
### 🔎 Search Terms

switch default

### 🕗 Version & Regression Information

The tsc complains that the the function return undefined if a add a default case, actually the default case will never be executed, just only for exhausive check.

### ⏯ Playground Link

https://www.typescriptlang.org/play?jsx=0&ts=5.4.4#code/C4TwDgpgBAyglgL2gXigcgDYEMBOBzCNKAH3QDsB7HAWywyNLQGdaN6T1q4y40AoAMYUyTYFCaIUUAN4BfKFiaxJg4aKgAjDBQEBrKKgA8AJQB8ACgBmALijmAlAdNRjj5M+l8o3qDgjAAVxwyKEsHPlk+VRExASYlVC0dXXMHWzIA6g0IHCcZLx8mAHc4YAEACzsJJEdPH3qoAUVoNEoaOjRbOoaev0DgqABGAp7Inu8mphbsfEIukfHvPqCQgCYF+rHxyZauHk78xfrlgYBaQagNny2enfQWOnp5o58TkIAGK+8bhoATCEsWACGGAzxejTUYjI6QgADccgZxCpFj9IrJ7EA

### 💻 Code

```ts
type Size = 'large' | 'normal' | 'small' | 'mini'
const size = {} as Size
const block = <R>(f: () => R) => {
    return f()
}

const css = block((): number => {
    switch (size) {
        case 'normal': {
            return 1
        }
        case 'large': {
            return 2
        }
        case 'mini': {
            return -1 
        }
        case 'small': {
            return 0
        }
        default: {
            const n: never = size
        }
    }
})
```

### 🙁 Actual behavior

The tsc think that the function of css block returns number | undefined.

### 🙂 Expected behavior

The tsc should know that the default case is never returned, and the function of css block should returns number

**Comment by RyanCavanaugh** [comment]
Not really sure how to reason about this one.

On the one hand, TS assumes that if you write code, you did so for a reason. So if there's a `default` case, then it must be reachable via some means that you know about but we don't, so it's fair game for computing the return type. You wouldn't be allowed to `return "foo";` here, for example, and the implicit `return undefined;` that exists at the implied exit point of the switch exists by the same logic.

On the other hand, in the absence of the `default`, we know the `switch` is exhaustive, so any code in the `default` block shouldn't need a `return`.

But again, you wrote the `default` block for a reason, and didn't `throw` out of it, so if that code isn't 100% dead, it ought to be an error. Otherwise you'll get an unsound `undefined` later in your program in a way that's really hard to diagnose.

I would just delete, or `return` or `throw` here for clarity. The current behavior doesn't seem *wrong* enough justify endangering people who do have an actually-reachable `default` (as is somewhat common in the case of e.g. side effects or array access)

**Comment by Mng12345** [comment]
The default case here is designed to allow the TypeScript compiler (tsc) to automatically detect when a new variant of the Size type is introduced. If the size in the default branch is of the new variant instead of never, it will trigger an error, reminding developers that a case has been missed.

**Comment by RyanCavanaugh** [comment]
You don't need one, though? https://www.typescriptlang.org/play?target=1#code/C4TwDgpgBAyglgL2gXigcgDYEMBOBzCNKAH3QDsB7HAWywyNLQGdaN6T1q4y40AoAMYUyTYFCaIUUAN4BfKFiaxJg4aKgAjDBQEBrKKgA8AJQA0APgAUAMwBcUSwEoD5qMefJX0vlF9QcEMAArjhkUNZOfLJ8qiJiAkxKqFo6upZO9mRB1BoQOC4yPn5MAO5wwAIAFg4SSM7efo1QAorQaJQ0dGj2DU19AcGhUACMRX3Rfb4tTG3Y+IQ9Y5O+AyFhAExLjROT021cPN2Fy42rQwC0w1Bbfju+0bKOQA

**Comment by Mng12345** [comment]
Yes, I can use the return type of the function as a workaround to perform exhaustive checks on the switch statement.

**Comment by Mng12345** [comment]
For absolute correctness, currently, the only way is to wrap the switch statement within a block function to ensure that no case is missed. Although typescript-eslint can also perform such checks, it's not always easy to configure typescript-eslint properly in every project.

**Comment by Mng12345** [comment]
It would be great if tsc can add an options config for switch exhausiveness check.
